### ISSUE TITLE
Contract based tests

### ISSUE THREAD

**Issue by jfiala** [swagger-api/swagger-codegen#4087]
##### Description

Currently there are no unittests ensuring that the contracts (i.e. the outcome of the language-generators) change only when they are supposed to change. What about adding at least one unittest for each language to secure this and to track all the generated output files (better their hashes)? For different configuration options of the language maybe additional tests make sense. I think this would also track breaking changes a lot easier...

**Comment by jfiala** [comment]
Automatically generating the output for all the petstore-variations and also allow for different language configurations would make the results more reliable. The current process is unreliable because if someone forgets to update the generated files they will not be validated. IMHO the expected outcome should be fixed, but the actual outcome should always get generated dynamically.

@wing328 can you please review the idea? If you want I can setup a separate module for this. It would also allow to replace the shellscripts with an integrated maven build.

**Comment by wing328** [comment]
@jfiala is your suggestion similar to https://github.com/civisanalytics/swagger-diff, which shows the differences between 2 Swagger specs?

**Comment by jfiala** [comment]
No, I think the manual process of running the shellscripts is tiresome and leads to outdated generated sample files (I've had this numerous times), which forces the updated to also fix other bugs that have creeped into the language. Also, the shellscripts use different contracts which adds additional complexity.

I'd suggest to setup a separate integration-test project which will  automatically generate all the languages including test variations with every build, forcing language updaters to be always consistent and otherwise the build will fail. 

Currently forgetting  to update the samples leads to no build errors and therefore bugs creep into the languages without getting noticed.

So my suggestion basically should be:
1. add a separate test module which generates all samples automatically using unittests and which will fail if the expected outcome differs from the actual (generated) outcome, forcing language updates to update the expected outcome and thus maintain the tests.

2. add additional tests for the test variations (e.g. security-contract petstore-securityy-test.yaml etc.).

3. maybe also implement cross-language consistency checks, as some features should be consistent throughout all languages

What do you think?

**Comment by jfiala** [comment]
@wing328 please take a look at the PR #4721, if you like it we can migrate all the shellscripts into unittests and also add the directory layout comparison to make sure everything is equal, maybe also improve the error reporting.

**Comment by jfiala** [comment]
@wing328 can you pls have a look at this suggestion to replace the shellscript and to enforce always up-to-date generated samples? Nearly every time I want to improve a language the samples are out of date and would cause compile errors if they are generated correctly. It is no fun having to generate the samples, fix errors due to outdated samples and then being able starting the real work. Just wondering why nobody else complains about this...

**Comment by jfiala** [comment]
@nickcmaynard  @cbornet can you pls also take a look/share opinion?

**Comment by wing328** [comment]
> Nearly every time I want to improve a language the samples are out of date and would cause compile errors if they are generated correctly.

@jfiala I agreed the outdated sample is something ignoring. 

The reason (why we didn't run run-all-petstore.sh to update all samples) is that from time to time, we add more test cases in the [test spec](https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/test/resources/2_0/petstore-with-fake-endpoints-models-for-testing.yaml) to cover more edge cases, we found the edge cases covered by the new test cases may or may not be ready for all generators. In other words, some generators are not ready for the new test cases (which is in some senses still a good thing as the test cases uncover issue that need to be fixed with the generators) and therefore CI tests will fail if updating the Petstore samples for all generators. (We did run run-all-petstore.sh at one point but stopped due to issues I explained above)

**Comment by jfiala** [comment]
@wing328 I agree up to some point, I think it should be a decision to opt out for a specific language for some time. But at least it should be a decision not something which happens because pom.xml is not overwritten or something like that. Additionally, the shellscripts are not suitable for the windows-based contributors, so this will be helpful anyway.

My suggestion for a roadmap:
1.) Port all shellscripts to Java Commands
2.) Add Unittest with complete directory structure comparison
3.) Opt out all the language unittest which fail and notify the maintainers that they should take a look when they have time.
4.) Add more unittests for e.g. snippets equal check to improve consistency between the languages
(e.g. beanvalidation-templates etc.) as centralized templates are not desired.
